sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala

```diff
@@ -251,7 +252,33 @@ trait ColumnResolutionHelper extends Logging with DataTypeErrorsBase {
     }
   }
 
-  def lookupVariable(nameParts: Seq[String]): Option[VariableReference] = {
+  sealed trait LookupVariableMode
+
+  /**
+   * Mode of variable lookup. Possible values:<p>
+   * DEFAULT => include local variables unless in EXECUTE IMMEDIATE.<p>
+   * INCLUDE_LOCAL_VARS => include local variables.<p>
+   * EXCLUDE_LOCAL_VARS => exclude local variables.<p>
+   */
+  object LookupVariableMode {
+    case object INCLUDE_LOCAL_VARS extends LookupVariableMode
+    case object EXCLUDE_LOCAL_VARS extends LookupVariableMode
+    case object DEFAULT extends LookupVariableMode
+  }
+
+  /**
+   * Look up variable by nameParts. If in SQL Script, first check local variables,
+   * if not found fall back to session variables,
+   * unless in EXECUTE IMMEDIATE or specified otherwise by sessionVariablesOnlyOpt.
+   * @param nameParts NameParts of the variable.
+   * @param mode DEFAULT => include local variables unless in EXECUTE IMMEDIATE.<p>
+   *             INCLUDE_LOCAL_VARS => include local variables.<p>
+   *             EXCLUDE_LOCAL_VARS => exclude local variables.<p>
+   * @return Reference to the variable.
+   */
+  def lookupVariable(
+      nameParts: Seq[String],
+      mode: LookupVariableMode = LookupVariableMode.DEFAULT): Option[VariableReference] = {
```

Comment:
looks like we can just check the local variable manager ThreadLocal and `AnalysisContext.isExecuteInmediate` here to decide the resolution behavior, instead of adding this parameter.